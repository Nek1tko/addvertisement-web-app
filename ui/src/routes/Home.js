import React, {useEffect} from "react";
import {FilterModal} from "../components/FilterModal";
import AdRecordsTableServer from "../components/AdRecordsTableServer";
import Box from "@mui/material/Box";
import {TextField} from "@mui/material";
import Button from "@material-ui/core/Button";
import SearchIcon from '@mui/icons-material/Search';
import axios from 'axios';
import authHeader from "../services/auth-header";
import AuthService from '../services/auth.service';
import {Redirect} from "react-router-dom";

const API_URL = "http://localhost:8080/ad/page";
const API_URL_COUNT = "http://localhost:8080/ad/page/count";

const HomeImpl = props => {
    const [searchValue, setSearchValue] = React.useState("");
    const [ads, setAds] = React.useState([]);
    const [rowCount, setRowCount] = React.useState(0);
    const [page, setPage] = React.useState(1);
    const {_, userId,} = AuthService.getUser();
    const [jsonFilter, setJsonFilter] = React.useState({});

    useEffect(() => {
        axios
            .post(API_URL_COUNT, {}, {headers: authHeader()})
            .then(res => {
                setRowCount(res.data);
            })
    }, []);

    useEffect(() => {
        axios
            .post(API_URL, {page: page}, {headers: authHeader()})
            .then(res => {
                setAds(res.data);
            })
    }, [page]);

    const handleClickSearch = () => {
        if (Object.keys(jsonFilter).length === 0) {
            axios
                .post(API_URL_COUNT, {title: searchValue}, {headers: authHeader()})
                .then(res => {
                    setRowCount(res.data);
                })

            axios
                .post(API_URL, {page: page, title: searchValue}, {headers: authHeader()})
                .then(res => {
                    setAds(res.data);
                })
        } else {
            axios
                .post(API_URL_COUNT, { ...jsonFilter, title: searchValue}, {headers: authHeader()})
                .then(res => {
                    setRowCount(res.data);
                })

            axios
                .post(API_URL, {...jsonFilter, title: searchValue}, {headers: authHeader()})
                .then(res => {
                    setAds(res.data);
                })
        }
    }

    return (
        <div>
            <Box sx={{display: "flex"}}>
                <Box sx={{width: 1 / 3, flex: 1}}>
                    <FilterModal page={page} setJsonFilter={setJsonFilter} setAds={setAds}/>
                </Box>
                <Box sx={{width: 1 / 3, flex: 1, marginTop: 3}}>
                    <TextField
                        id="searchTextField"
                        type="search"
                        size="small"
                        label="Поиск"
                        variant="outlined"
                        onChange={e => {
                            setSearchValue(e.target.value);
                        }}
                        value={searchValue}
                        fullWidth
                    />
                </Box>
                <Box sx={{width: 1 / 3, flex: 1, marginTop: 3}}>
                    <Button
                        id="searchButton"
                        variant="contained"
                        onClick={handleClickSearch}
                        endIcon={<SearchIcon/>}
                        color="primary"
                    >
                        Поиск
                    </Button>
                </Box>
            </Box>
            <AdRecordsTableServer
                userId={userId}
                ads={ads}
                rowCount={rowCount}
                API_URL={API_URL}
                setPage={(page) => setPage(page)}
                setAds={(ads) => setAds(ads)}
                history={props.history}
            />
        </div>
    );
}

const Home = props => {
    if (!AuthService.getUser()) {
        return <Redirect to="/login"/>
    }

    return <HomeImpl history={props.history}/>
};

export default Home;