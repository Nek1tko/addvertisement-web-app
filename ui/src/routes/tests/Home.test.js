/**
 * @group unit
 */

import AuthService from "../../services/auth.service";
import axios from "axios";
import { createBrowserHistory } from "history";
import React from "react";
import { mount } from "enzyme";
import { Router } from "react-router-dom";
import Home from "../Home";
import { act } from "react-dom/test-utils";
import authHeader from "../../services/auth-header";

jest.mock("axios");
jest.mock("../../services/auth.service");
jest.mock("../../services/auth-header");

const eventSearch = {
    target: {
        value: "mockName"
    }
};

describe('HomeTests', () => {
    let wrapper, props, useStateSpy;
    const setState = jest.fn();

    beforeEach(() => {
        props = {
            history: {
                ...createBrowserHistory(),
                push: jest.fn()
            }
        };
        AuthService.getUser.mockReturnValue({ user: "mockUser" });
        axios.post.mockResolvedValue({
            data: [
                {
                    id: 1,
                    name: "mock name"
                },
                {
                    id: 2,
                    name: "mock name"
                },
            ]
        });
        axios.get.mockResolvedValue({
            data: [
                {
                    id: 1,
                    name: "mock name"
                },
                {
                    id: 2,
                    name: "mock name"
                },
            ]
        });
        authHeader.mockReturnValue({ header: "mockHeader" });
        useStateSpy = jest.spyOn(React, 'useState');
        useStateSpy.mockImplementation((state) => [state, setState]);
        act(async () => wrapper = await mount(
            <Router history={props.history}>
                <Home {...props} />
            </Router>));
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    test('HomeRenderTest', async () => {
        await act(async () => {
            await mount(<Router history={props.history}>
                <Home {...props} />
            </Router>);
        });
    });

    test('HomeSnapshotTest', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('HomeSearchButtonNegativeTest', async () => {
        await act(async () => {
            await wrapper.find('#searchButton').first().props().onClick();
        });
        expect(setState).toHaveBeenCalledTimes(11);
    });

    test('HomeSearchButtonNegativeTest', async () => {
        await act(async () => {
            await wrapper.find('#searchTextField').first().props().onChange(eventSearch);
        });
        expect(setState).toHaveBeenCalledTimes(10);
    });

    test('HomeNotAuthorizedTest', () => {
        AuthService.getUser.mockReturnValue(null);
        wrapper = mount(
            <Router history={props.history}>
                <Home {...props} />
            </Router>);
    });
});
