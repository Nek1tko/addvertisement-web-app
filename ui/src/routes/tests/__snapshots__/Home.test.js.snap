// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeTests HomeSnapshotTest 1`] = `ReactWrapper {}`;
